moda-contracts/src/Marketplace.sol

```diff
@@ -0,0 +1,236 @@
+import "../src/interfaces/IMarketplace.sol";
+import "../src/interfaces/Releases/IReleasesRegistration.sol";
+
+/**
+ * @title Marketplace
+ * @dev This contract allows buying and selling of Releases and charges a fee on each sale.
+ */
+contract Marketplace is IMarketplace, ERC1155Holder, ReentrancyGuard, AccessControl {
+    using SafeERC20 for IERC20;
+
+    // State Variables
+
+    address _usdc;
+    address _catalog;
+    address payable public treasury;
+    uint256 public treasuryFee;
+
+    /// @dev seller => saleId => Sale
+    mapping(address => Sale[]) private sales;
+
+    // Errors
+    error CannotBeZeroAddress();
+    error TreasuryFeeCannotBeZero();
+    error TokenAmountCannotBeZero();
+    error MaxCountCannotBeZero();
+    error StartCannotBeAfterEnd(uint256 startTime, uint256 endTime);
+    error InsufficientSupply(uint256 remainingSupply);
+    error MaxSupplyReached(uint256 maxSupplyPerWallet);
+    error SaleNotStarted(uint256 startTime, uint256 currentTime);
+    error SaleHasEnded(uint256 endTime, uint256 currentTime);
+    error ReleasesIsNotRegistered();
+
+    /**
+     * @dev Constructor
+     * @param treasury_ - The address of the organizations treasury
+     * @param treasuryFee_ - The percentage that will be transferred
+     * to the treasury on each sale. Based on a denominator of 10_000 e.g. 1000 = 10%
+     * @param usdc - The USDC token address
+     * @param catalog - The address of the catalog contract
+     */
+    constructor(address payable treasury_, uint256 treasuryFee_, address usdc, address catalog) {
+        if (usdc == address(0)) revert CannotBeZeroAddress();
+        if (treasuryFee_ == 0) revert TreasuryFeeCannotBeZero();
+        if (treasury_ == address(0)) revert CannotBeZeroAddress();
+        treasury = treasury_;
+        treasuryFee = treasuryFee_;
+        _usdc = usdc;
+        _catalog = catalog;
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // External Functions
+
+    /**
+     * @inheritdoc IMarketplace
+     */
+    function createSale(
+        address payable beneficiary,
+        address releases,
+        uint256 tokenId,
+        uint256 amountTotal,
+        uint256 pricePerToken,
+        uint256 startAt,
+        uint256 endAt,
+        uint256 maxCountPerWallet
+    ) external {
+        if (IReleasesRegistration(_catalog).getReleasesOwner(releases) == address(0)) {
+            revert ReleasesIsNotRegistered();
+        }
+        if (beneficiary == address(0)) revert CannotBeZeroAddress();
+        if (amountTotal == 0) revert TokenAmountCannotBeZero();
+        if (endAt != 0 && startAt > endAt) {
+            revert StartCannotBeAfterEnd(startAt, endAt);
+        }
+        if (maxCountPerWallet == 0) revert MaxCountCannotBeZero();
+        IERC1155(releases).safeTransferFrom(_msgSender(), address(this), tokenId, amountTotal, "");
+
+        sales[_msgSender()].push(
+            Sale({
+                releaseOwner: _msgSender(),
```

Comment:
`releasesOwner` is not necessarily the owner. It could be a manager. Do we need to store this? If an artist creates a sale can their manager come back and remove the sale? We need to allow any eligible party to be able to mutate a sale listing